table/tables/tables.go

```diff
@@ -468,12 +469,19 @@ func (t *tableCommon) AddRecord(ctx sessionctx.Context, r []types.Datum, skipHan
 
 	for _, col := range t.WritableCols() {
 		var value types.Datum
-		if col.State != model.StatePublic {
+		if col.State != model.StatePublic && isInsert {
 			// If col is in write only or write reorganization state, we must add it with its default value.
 			value, err = table.GetColOriginDefaultValue(ctx, col.ToInfo())
 			if err != nil {
 				return 0, errors.Trace(err)
 			}
+			// add value to `r` for dirty db in transaction.
+			// Otherwise when update will panic cause by get value of column in write only state from dirty db.
+			if col.Offset < len(r) {
```

Comment:
```
create table t (i int);
insert into t (i, _tidb_rowid) values (1, 2); -- alter table t add column j int after i; 
```
col.Offset is 1 becaues it is "j", and len(r) is 2 due to it appends _tidb_rowid at the end, then r[1] will be reset?

Reply:
ye, but this doesn't affect the insert value , because after will use `row` to encode value, and `r` will only use in the dirty table now.

Reply:
so if 
```
begin;
insert into t (i, _tidb_rowid) values (1, 2); -- alter table t add column j int default 3 after i; 
select _tidb_rowid from t where i = 1;
```
will get 3 not 2?

Reply:
will get 2.

Reply:
dirty table use `handle` as key, `row` record as value, So It will be ok.

Reply:
OK